### DISCUSSION TITLE
Static File route intercept?

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#2429] [Q&A]
I'm digging the new static file support with mysite.com/files/somefile.pdf serving up files as a clean URL.  We have a a need to track the files that are downloaded and I'm curious if there is a way to hook into the routes and fire some custom code.  Our current plan is to make an API endpoint like mysite.com/api/download/somefile.pdf to capture the telemetry (custom logging) and then return a redirect to the actual file, but if there is a fun way to hook into the routes, I'd love to hear it!

**Comment by leigh-pointer** [comment]
Hi @markdav-is this was discussed here https://github.com/oqtane/oqtane.framework/discussions/2261 . I haven't actioned it.

**Reply by sbwalker** [reply]
A server-side event would probably be the best way to add support for this in the near term - which would allow developers to include their own logic OnDownload.

**Comment by sbwalker** [comment]
@markdav-is #2449 allows developers to consume events raised by the core framework. It uses the standard the .NET EventHandler capability in conjunction with the SyncManager to raise SyncEvents (which inherit from the standard EventArgs). 

The tricky part with event handling in a web context is that most services are short-lived (ie. either scoped or transient) - which is a problem because in order for you to be able to capture an event, you need to have a service which is guaranteed to be running at the precise moment when the event is fired. This is where IHostedServices can be very helpful. IHostedServices startup at the same time as your web application and run on a background thread for the lifetime of your application. This makes them a perfect candidate for capturing events. The challenge with IHostedServices are that they do not support standard dependency injection for scoped services so you need to do some extra work to instantiate those services.

The following is an example of how to capture the File Download event. You will notice that it registers the EntityChanged event handler and then inspects the SyncEvent to determine if the Entity and Action match the specific event you are interested in. You should also note that due to the multi-tenant nature of Oqtane it is important to set the tenant before calling any repository methods.

```
    public class FileDownloadHostedService : IHostedService
    {
        private readonly IServiceScopeFactory _serviceScopeFactory;
        private readonly ISyncManager _syncManager;

        public FileDownloadHostedService(IServiceScopeFactory serviceScopeFactory, ISyncManager syncManager)
        {
            _serviceScopeFactory = serviceScopeFactory;
            _syncManager = syncManager;
        }

        void EntityChanged(object sender, SyncEvent e)
        {
            // consume SyncEvent
            if (e.EntityName == "File" && e.Action == "Download")
            {
                using (var scope = _serviceScopeFactory.CreateScope())
                {
                    // create scoped services
                    var tenantManager = scope.ServiceProvider.GetRequiredService<ITenantManager>();
                    var fileRepository = scope.ServiceProvider.GetRequiredService<IFileRepository>();

                    // set tenant
                    tenantManager.SetTenant(e.TenantId);

                    // get file and do whatever you need
                    var file = fileRepository.GetFile(e.EntityId);
                }
            }
        }

        public Task StartAsync(CancellationToken cancellationToken)
        {
            _syncManager.EntityChanged += EntityChanged;

            return Task.CompletedTask;
        }

        public Task StopAsync(CancellationToken cancellationToken)
        {
            return Task.CompletedTask;
        }

        public void Dispose()
        {
            _syncManager.EntityChanged -= EntityChanged;
        }
    }
```
All of the core framework entities have been enhanced to raise Add, Update, and Delete events in this manner - which opens up many new opportunities for developers to extend the framework server-side.

Note that SyncManager already existed in prior versions of Oqtane but was only used to synchronize server-side events with client-side state. It still serves this purpose as well (with some Refresh and Reload actions that were introduced for this specific scenario).

**Reply by sbwalker** [reply]
I should also mention that the File Download event is only raised on "download" events - not for standard file serving scenarios. If you are using the new file handler introduced in 3.2.0 you need to remember to include the ?download QueryString parameter on the Url so that the download logic is triggered.

**Comment by leigh-pointer** [comment]
@sbwalker this is really great stuff.  This will certainly open up a stream of opportunities. 🚀
